### PR TITLE
Adding property - fabric forwarding mode anycast-gateway, to Cisco_Interface

### PR DESCRIPTION
**Pull request by smigopal**
# Description:

Supporting property fabric forwarding mode anycast-gateway under Interface
# Minitests:

---

rtpfe1@agent-lab11-ws:~/smigopal/feb_interface/cisco-network-node-utils$ ruby tests/test_interface.rb -v -n test_interface_fabric_forwarding_anycast_gateway -- 10.122.197.125 admin admin
Run options: -v -n test_interface_fabric_forwarding_anycast_gateway -- --seed 26306

Running:

Node under test:
- name  - agent-lab11-nx
- type  - N9K-NXOSV
- image - bootflash:///nxos.7.0.3.I2.1.bin

conf : ["interface vlan11"]
conf : [" fabric forwarding anycast-gateway-mac 1223.3445.5668"]
conf : ["interface vlan11", "no fabric forwarding mode anycast-gateway"]
conf : ["interface vlan11", " fabric forwarding mode anycast-gateway"]
conf : ["interface vlan11", "no fabric forwarding mode anycast-gateway"]
conf : ["no fabric forwarding anycast-gateway-mac "]
conf : ["interface vlan11", " fabric forwarding mode anycast-gateway"]
conf : ["feature fabric forwarding"]
conf : ["interface vlan11", " fabric forwarding mode anycast-gateway"]
conf : ["interface ethernet1/1"]
conf : ["interface ethernet1/1", " fabric forwarding mode anycast-gateway"]
TestInterface#test_interface_fabric_forwarding_anycast_gateway = 10.29 s = .

Finished in 10.289229s, 0.0972 runs/s, 0.5831 assertions/s.

1 runs, 6 assertions, 0 failures, 0 errors, 0 skips
Coverage report generated for MiniTest to /home/rtpfe1/smigopal/feb_interface/cisco-network-node-utils/coverage. 669 / 1287 LOC (51.98%) covered.
rtpfe1@agent-lab11-ws:~/smigopal/feb_interface/cisco-network-node-utils$
# Rubocop:

---

rtpfe1@agent-lab11-ws:~/smigopal/feb_interface/cisco-network-node-utils$ rubocop tests/test_interface.rb
warning: parser/current is loading parser/ruby22, which recognizes
warning: 2.2.3-compliant syntax, but you are running 2.2.1.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
Inspecting 1 file
.

1 file inspected, no offenses detected
rtpfe1@agent-lab11-ws:~/smigopal/feb_interface/cisco-network-node-utils$ rubocop lib/cisco_node_utils/interface.rb
warning: parser/current is loading parser/ruby22, which recognizes
warning: 2.2.3-compliant syntax, but you are running 2.2.1.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
Inspecting 1 file
.

1 file inspected, no offenses detected
rtpfe1@agent-lab11-ws:~/smigopal/feb_interface/cisco-network-node-utils$

### CODE REVIEW

File: lib/cisco_node_utils/interface.rb [hunk 1]
```diff
@@ -205,6 +206,32 @@ def fabricpath_feature_set(fabricpath_set)
...
+      config_get('interface', 'fabric_forwarding_anycast_gateway', @name)
+    end
+
+    def fabric_forwarding_anycast_gateway=(state)
+      begin
+        Feature.fabric_forwarding_enable
+        no_cmd = (state ? '' : 'no')
+        config_set('interface',
+                   'fabric_forwarding_anycast_gateway', @name, no_cmd)
+        expected_state = state
```
Comment (mikewiebe) [inline lib/cisco_node_utils/interface.rb]: This line `expected_state = state` is unnecessary.  Just use `state`.

File: lib/cisco_node_utils/cmd_ref/interface.yaml [hunk 2]
```diff
@@ -46,6 +46,12 @@ encapsulation_dot1q:
   config_set_append: "%s encapsulation dot1q %s"
   default_value: ""
```
Comment (mikewiebe) [inline lib/cisco_node_utils/cmd_ref/interface.yaml:47]: Please update the CHANGELOG to include this property.

File: tests/test_interface.rb [hunk 3]
```diff
@@ -958,6 +959,44 @@ def test_interface_ipv4_arp_timeout
     assert_raises(RuntimeError) { nonvlanint.ipv4_arp_timeout = 300 }
   end
 
+  def test_interface_fabric_forwarding_anycast_gateway
+    # Setup
+    config('no interface vlan11')
```
Comment (mikewiebe) [inline tests/test_interface.rb:964]: Instead of using the `config` api, please use the `Interface` and `VxlanGlobal` providers to accomplish the same task.

Update: Thinking more... using the `config` api to remove the vlan is fine.

File: lib/cisco_node_utils/interface.rb [hunk 4]
```diff
@@ -205,6 +206,32 @@ def fabricpath_feature_set(fabricpath_set)
...
+        fail if fabric_forwarding_anycast_gateway.to_s != expected_state.to_s
+      end
+    rescue Cisco::CliError => e
+      vxlan_global = VxlanGlobal.new
+      if vxlan_global.anycast_gateway_mac == ''
```
Comment (mikewiebe) [inline lib/cisco_node_utils/interface.rb]: You can collapse these two lines into one like this...

`if VxlanGlobal.new.anycast_gateway_mac == ''`

File: tests/test_interface.rb [hunk 5]
```diff
@@ -958,6 +959,44 @@ def test_interface_ipv4_arp_timeout
...
+    int = Interface.new('vlan11')
+    config('feature fabric forwarding')
+    foo = VxlanGlobal.new
+    foo.anycast_gateway_mac = '1223.3445.5668'
+
+    # 1. Testing default
+    int.fabric_forwarding_anycast_gateway = false
+    assert_equal(int.default_fabric_forwarding_anycast_gateway,
```
Comment (mikewiebe) [inline tests/test_interface.rb:972]: Please use the more specific `refute` method in this case.

https://github.com/cisco/cisco-network-node-utils/blob/release_1.2.0/docs/README-develop-best-practices.md#mt2

File: tests/test_interface.rb [hunk 6]
```diff
@@ -958,6 +959,44 @@ def test_interface_ipv4_arp_timeout
...
+                 int.fabric_forwarding_anycast_gateway)
+
+    # 2. Testing non-default
+    int.fabric_forwarding_anycast_gateway = true
+    assert_equal(true, int.fabric_forwarding_anycast_gateway)
```
Comment (mikewiebe) [inline tests/test_interface.rb]: Please use the more specific `assert` method in this case.

https://github.com/cisco/cisco-network-node-utils/blob/release_1.2.0/docs/README-develop-best-practices.md#mt2

File: tests/test_interface.rb [hunk 7]
```diff
@@ -958,6 +959,44 @@ def test_interface_ipv4_arp_timeout
...
+
+    # 3. Setting back to default
+    int.fabric_forwarding_anycast_gateway =
+      int.default_fabric_forwarding_anycast_gateway
+    assert_equal(int.default_fabric_forwarding_anycast_gateway,
```
Comment (mikewiebe) [inline tests/test_interface.rb]: `refute` again here.

File: tests/test_interface.rb [hunk 8]
```diff
@@ -958,6 +959,44 @@ def test_interface_ipv4_arp_timeout
...
+                 int.fabric_forwarding_anycast_gateway)
+
+    # 4. Removing fabric forwarding anycast gateway mac
+    foo.anycast_gateway_mac = foo.default_anycast_gateway_mac
+    assert_raises(RuntimeError) { int.fabric_forwarding_anycast_gateway = true }
```
Comment (mikewiebe) [inline tests/test_interface.rb:985]: Presumably we already have similar tests for `VxlanGlobal`?

**Comment by mikewiebe** [comment]
@smigopal, @glennmatthews  is refactoring VxlanGlobal a bit which will result in a few changes to this code as well.  He should have a PR out soon, but let's wait on that before we continue with this code review.

**Comment by glennmatthews** [comment]
See #220

File: lib/cisco_node_utils/interface.rb [hunk 9]
```diff
@@ -205,6 +206,32 @@ def fabricpath_feature_set(fabricpath_set)
       FabricpathGlobal.fabricpath_feature_set(fabricpath_set)
     end
 
+    def fabric_forwarding_anycast_gateway
+      config_get('interface', 'fabric_forwarding_anycast_gateway', @name)
+    end
+
+    def fabric_forwarding_anycast_gateway=(state)
+      begin
```
Comment (glennmatthews) [inline lib/cisco_node_utils/interface.rb]: `begin`...`end` is unnecessary here.

File: lib/cisco_node_utils/interface.rb (hunk 4 above)
Comment (glennmatthews) [inline lib/cisco_node_utils/interface.rb]: With the change in #220 this will be:

``` ruby
rescue Cisco::CliError => e
  anycast_gateway_mac = OverlayGlobal.new.anycast_gateway_mac
  if anycast_gateway_mac.nil? || # nv overlay evpn is disabled
     anycast_gateway_mac.empty? # nv overlay evpn is enabled but mac not set
    raise ...
```

File: tests/test_interface.rb [hunk 10]
```diff
@@ -958,6 +959,44 @@ def test_interface_ipv4_arp_timeout
     assert_raises(RuntimeError) { nonvlanint.ipv4_arp_timeout = 300 }
   end
 
+  def test_interface_fabric_forwarding_anycast_gateway
+    # Setup
+    config('no interface vlan11')
+    int = Interface.new('vlan11')
+    config('feature fabric forwarding')
+    foo = VxlanGlobal.new
+    foo.anycast_gateway_mac = '1223.3445.5668'
```
Comment (glennmatthews) [inline tests/test_interface.rb:967]: Explicitly configuring 'feature fabric forwarding' isn't needed for VxlanGlobal/OverlayGlobal - unless it's needed for the interface property, please remove it. These three lines should just become:

```
OverlayGlobal.new.anycast_gateway_mac = '1223.3445.5668'
```

File: tests/test_interface.rb (hunk 5 above)
Comment (glennmatthews) [inline tests/test_interface.rb:972]: Eh? assert_equal is the correct thing to use here since you're comparing against the default value.

However, the line above should be:

``` ruby
int.fabric_forwarding_anycast_gateway = int.default_fabric_forwarding_anycast_gateway
```

File: tests/test_interface.rb (hunk 8 above)
Comment (glennmatthews) [inline tests/test_interface.rb:985]: Sure, but this is testing the specific error case in the interface code.

File: tests/test_interface.rb (hunk 5 above)
Comment (mikewiebe) [inline tests/test_interface.rb:972]: True, my mistake.

File: tests/test_interface.rb (hunk 8 above)
Comment (mikewiebe) [inline tests/test_interface.rb:985]: Ok, just making sure we don't have any unnecessary overlap.

File: tests/test_interface.rb (hunk 10 above)
Comment (mikewiebe) [inline tests/test_interface.rb:967]: @glennmatthews It is needed in `I2` images.

```
agent-lab10-nx(config)# sh run | i 'feature fabric'
agent-lab10-nx(config)# conf t
agent-lab10-nx(config)# int vlan 11
agent-lab10-nx(config-if)# fabric fo?
                           ^
% Invalid command at '^' marker.
agent-lab10-nx(config-if)# 
agent-lab10-nx(config-if)# feature fabric forwarding
agent-lab10-nx(config)# int vlan 11
agent-lab10-nx(config-if)# fabric forwarding mode anycast-gateway 
```

vs. the behavior on `I3` images...

```
dt-n9k5-1(config)# feature interface-vlan 
dt-n9k5-1(config)# 
dt-n9k5-1(config)# sh run | i 'feature fabric'
dt-n9k5-1(config)# 
dt-n9k5-1(config)# int vlan 11
dt-n9k5-1(config-if)# fabric forwarding mode anycast-gateway 
dt-n9k5-1(config-if)# 
```

Comment (mikewiebe) [inline tests/test_interface.rb:967]: @smigopal I spoke with @glennmatthews and he clarified that his point is that `OverlayGlobal.new.anycast_gateway_mac` will handle turning `feature fabric forwarding` on `I2` images and ignore cli failures on `I3` images.

Comment (smigopal) [inline tests/test_interface.rb:967]: @mikewiebe : Awesome. Just to verify, this means I wouldn't need to enable the feature using "config('feature fabric forwarding')". Just the setter should suffice = "OverlayGlobal.new.anycast_gateway_mac = '1223.3445.5668'"

Comment (mikewiebe) [inline tests/test_interface.rb:967]: That's correct.

File: lib/cisco_node_utils/interface.rb (hunk 1 above)
Comment (smigopal) [inline lib/cisco_node_utils/interface.rb]: Done.

File: lib/cisco_node_utils/cmd_ref/interface.yaml (hunk 2 above)
Comment (smigopal) [inline lib/cisco_node_utils/cmd_ref/interface.yaml:47]: Done

File: tests/test_interface.rb (hunk 3 above)
Comment (smigopal) [inline tests/test_interface.rb:964]: Ignored

File: lib/cisco_node_utils/interface.rb (hunk 9 above)
Comment (smigopal) [inline lib/cisco_node_utils/interface.rb]: Done

File: tests/test_interface.rb (hunk 6 above)
Comment (smigopal) [inline tests/test_interface.rb]: Done

File: tests/test_interface.rb (hunk 5 above)
Comment (smigopal) [inline tests/test_interface.rb:972]: Done

File: lib/cisco_node_utils/interface.rb (hunk 4 above)
Comment (smigopal) [inline lib/cisco_node_utils/interface.rb]: Done

File: tests/test_interface.rb (hunk 8 above)
Comment (smigopal) [inline tests/test_interface.rb:985]: Ignored

File: tests/test_interface.rb (hunk 10 above)
Comment (smigopal) [inline tests/test_interface.rb:967]: Ignored.

File: lib/cisco_node_utils/overlay_global.rb [hunk 11]
```diff
@@ -131,7 +131,7 @@ def anycast_gateway_mac
     def anycast_gateway_mac=(mac_addr)
       fail TypeError unless mac_addr.is_a?(String)
 
-      Feature.nv_overlay_evpn_enable unless Feature.nv_overlay_evpn_enabled?
+      Feature.nv_overlay_evpn_enable
```
Comment (glennmatthews) [inline lib/cisco_node_utils/overlay_global.rb:134]: Why this change?

File: tests/test_interface.rb [hunk 12]
```diff
@@ -958,6 +959,42 @@ def test_interface_ipv4_arp_timeout
...
+  def test_interface_fabric_forwarding_anycast_gateway
+    # Setup
+    config('no interface vlan11')
+    int = Interface.new('vlan11')
+    foo = OverlayGlobal.new
+    foo.anycast_gateway_mac = '1223.3445.5668'
+
+    # 1. Testing default
+    int.fabric_forwarding_anycast_gateway =
+      int.default_fabric_forwarding_anycast_gateway
```
Comment (glennmatthews) [inline tests/test_interface.rb:971]: I'd prefer removing these two lines - have the test below be for verifying that the default state of a newly created Vlan matches the expected default value, rather than explicitly setting it to default.
